dashboard/cybershake_progress.py

```diff
@@ -0,0 +1,373 @@
+    progress_df["total", COMPLETED_R_COUNT_COL] = 0
+    progress_df["total", R_COUNT_COL] = r_counts
+
+    # Get actual core hours for all faults, assumes faults are run
+    # in alphabetical order
+    chours_df = get_chours_used(faults)
+
+    # Add actual data to progress df
+    total = np.nan
+    for proc_type in PROCESS_TYPES:
+        values = chours_df[proc_type]
+        progress_df[proc_type, ACT_CORE_HOURS_COL] = values
+        total = values if total is np.nan else total + values
+    progress_df["total", ACT_CORE_HOURS_COL] = total
+    progress_df["total", COMPLETED_R_COUNT_COL] = chours_df[COMPLETED_R_COUNT_COL]
+    progress_df["total", MISSING_DATA_FLAG_COL] = chours_df[MISSING_DATA_FLAG_COL]
+
+    return progress_df
+
+
+def load_progress_df(file: str):
+    return pd.read_csv(file, index_col=[0], header=[0, 1])
+
+
+def print_progress(progress_df: pd.DataFrame, cur_fault_ix: int = None):
+    """Prints the progress dataframe in a nice format.
+
+    If cur_fault_ix is specified then all uncompleted faults are printed and
+    the latest completed fault with the previous and next five faults.
+
+    Otherwise all faults are printed.
+    """
+
+    def get_usage_str(fault: str, proc_type: str):
+        return "{:.2f}/{:.2f}".format(
+            progress_df.loc[fault, (proc_type, ACT_CORE_HOURS_COL)],
+            progress_df.loc[fault, (proc_type, EST_CORE_HOURS_COL)],
+        )
+
+    if cur_fault_ix is not None:
+        to_print_mask = (
+            progress_df["total", COMPLETED_R_COUNT_COL].values
+            != progress_df["total", R_COUNT_COL].values
+        ) & (progress_df["total", COMPLETED_R_COUNT_COL].values > 0)
+
+        to_print_mask[cur_fault_ix - 5 : cur_fault_ix + 6] = True
+    else:
+        to_print_mask = np.ones(progress_df.shape[0], dtype=bool)
+
+    template_fmt = "{:<15}{:<12}{:<12}{:<12}{:<12}{:<20}{:<12}"
+    print(
+        template_fmt.format(
+            "Fault name",
+            "EMOD3D",
+            "HF",
+            "BB",
+            "Total",
+            "Comp. realisations",
+            "Missing metadata",
+        )
+    )
+    for fault in progress_df.index.values[to_print_mask]:
+        print(
+            template_fmt.format(
+                fault,
+                get_usage_str(fault, const.ProcessType.EMOD3D.str_value),
+                get_usage_str(fault, const.ProcessType.HF.str_value),
+                get_usage_str(fault, const.ProcessType.BB.str_value),
+                get_usage_str(fault, "total"),
+                "{}/{}".format(
+                    progress_df.loc[fault, ("total", COMPLETED_R_COUNT_COL)],
+                    progress_df.loc[fault, ("total", R_COUNT_COL)],
+                ),
+                "True"
+                if progress_df.loc[fault, ("total", MISSING_DATA_FLAG_COL)]
+                else "False",
+            )
+        )
+
+    lf_act = progress_df[const.ProcessType.EMOD3D.str_value, ACT_CORE_HOURS_COL].sum()
```

Comment:
It almost seems nicer if we loop it. But that has disadvantages too.

Reply:
Good catch, no idea why I did it that way, its rather ugly....